apm-sniffer/bytebuddy-patch/src/test/java/org/apache/skywalking/apm/agent/bytebuddy/SWClassFileLocator.java

```diff
@@ -0,0 +1,174 @@
+
+import java.io.IOException;
+import java.lang.instrument.Instrumentation;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Resolve auxiliary type from Instrumentation.getAllLoadedClasses().
+ * Get class bytecode from separate thread to bypass jdk limitation or bug: https://github.com/raphw/byte-buddy/issues/1434
+ */
+public class SWClassFileLocator implements ClassFileLocator {
+    private static ILog LOGGER = LogManager.getLogger(SWClassFileLocator.class);
+
+    private final ForInstrumentation.ClassLoadingDelegate classLoadingDelegate;
+    private Instrumentation instrumentation;
+    private ClassLoader classLoader;
+    private String[] typeNameTraits = {"auxiliary$", "ByteBuddy$", "sw$"};
+    private BlockingQueue<ResolutionFutureTask> queue = new LinkedBlockingDeque<>();
+    private Thread thread;
+    private int timeoutSeconds = 2;
+    private volatile boolean closed;
+
+    public SWClassFileLocator(Instrumentation instrumentation, ClassLoader classLoader) {
+        this.instrumentation = instrumentation;
+        this.classLoader = classLoader;
+        classLoadingDelegate = ForInstrumentation.ClassLoadingDelegate.ForDelegatingClassLoader.of(classLoader);
+
+        // Use thread instead of ExecutorService here, avoiding conflicts with apm-jdk-threadpool-plugin
+        thread = new Thread(() -> {
+            while (!closed) {
+                try {
+                    ResolutionFutureTask task = queue.poll(5, TimeUnit.SECONDS);
+                    if (task != null) {
+                        try {
+                            Resolution resolution = getResolution(task.getClassName());
+                            task.getFuture().complete(resolution);
+                        } catch (Throwable e) {
+                            task.getFuture().completeExceptionally(e);
+                        }
+                    }
+                } catch (InterruptedException e) {
+                    // ignore interrupted error
+                } catch (Throwable e) {
+                    LOGGER.error(e, "Resolve bytecode of class failure");
+                }
+            }
+        }, "SWClassFileLocator");
+        thread.setDaemon(true);
+        thread.start();
+    }
+
+    @Override
+    public Resolution locate(String name) throws IOException {
+        if (!match(name)) {
+            return new Resolution.Illegal(name);
+        }
+        if (closed) {
+            throw new IOException("resolve class failure: closed");
+        }
+        // get class binary representation in a clean thread, avoiding nest calling transformer!
+        ResolutionFutureTask futureTask = new ResolutionFutureTask(name);
+        queue.offer(futureTask);
+        try {
+            return futureTask.getFuture().get(timeoutSeconds, TimeUnit.SECONDS);
```

Comment:
offer in 2 sec, but `queue.poll` in 5 sec, what is the logic behind this?

Reply:
Default resolve timeout is 2 sec,  poll timeout has nothing to do with it, just to avoid being blocked for a long time.

Reply:
OK, so, you just want the `poll` would be not blocked if there is nothing in the queue, and still be able to exit due to `closed`. Am I right?

Reply:
Yes, you are right.